hw/mcu/nordic/nrf5340/src/hal_gpio.c

```diff
@@ -0,0 +1,563 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <stdlib.h>
+#include <assert.h>
+#include "os/mynewt.h"
+#include "hal/hal_gpio.h"
+#include "mcu/cmsis_nvic.h"
+#include "nrf.h"
+#include "mcu/nrf5340_hal.h"
+
+/* XXX:
+ * 1) The code probably does not handle "re-purposing" gpio very well.
+ * "Re-purposing" means changing a gpio from input to output, or calling
+ * gpio_init_in and expecting previously enabled interrupts to be stopped.
+ *
+ */
+
+/* GPIO interrupts */
+#define HAL_GPIO_MAX_IRQ        8
+
+#if MYNEWT_VAL(MCU_GPIO_USE_PORT_EVENT)
+#define HAL_GPIO_SENSE_TRIG_NONE    0x00 /* just 0 */
+#define HAL_GPIO_SENSE_TRIG_BOTH    0x01 /* something else than both below */
+#define HAL_GPIO_SENSE_TRIG_HIGH    0x02 /* GPIO_PIN_CNF_SENSE_High */
+#define HAL_GPIO_SENSE_TRIG_LOW     0x03 /* GPIO_PIN_CNF_SENSE_Low */
+#endif
+
+/* Storage for GPIO callbacks. */
+struct hal_gpio_irq {
+    hal_gpio_irq_handler_t func;
+    void *arg;
+#if MYNEWT_VAL(MCU_GPIO_USE_PORT_EVENT)
+    int pin;
+    uint8_t sense_trig;
+#endif
+};
+
+static struct hal_gpio_irq hal_gpio_irqs[HAL_GPIO_MAX_IRQ];
+
+/**
+ * gpio init in
```

Comment:
First line of comment just duplicates name of the function with hal and underscores removed. It does not provide any useful information.